modules/ppcp-axo/src/Gateway/AxoGateway.php

```diff
@@ -277,6 +322,100 @@ public function process_payment( $order_id ) {
+	 *
+	 * @param WC_Order $wc_order The WooCommerce order.
+	 * @param string   $token    The PayPal order token.
+	 *
+	 * @return array
+	 */
+	protected function process_3ds_return( WC_Order $wc_order, string $token ) : array {
+		try {
+			$paypal_order = $this->order_endpoint->order( $token );
+
+			if ( ! $paypal_order->status()->is( OrderStatus::COMPLETED ) ) {
+				return array(
+					'result'  => 'failure',
+					'message' => __( '3D Secure authentication was not completed successfully. Please try again.', 'woocommerce-paypal-payments' ),
+				);
+			}
+
+			/**
+			 * This filter controls if the method 'process()' from OrderProcessor will be called.
+			 * So you can implement your own for example on subscriptions
+			 *
+			 * - true bool controls execution of 'OrderProcessor::process()'
+			 * - $this \WC_Payment_Gateway
+			 * - $wc_order \WC_Order
+			 */
+			$process = apply_filters( 'woocommerce_paypal_payments_before_order_process', true, $this, $wc_order );
+			if ( $process ) {
+				$this->order_processor->process_captured_and_authorized( $wc_order, $paypal_order );
+			}
+		} catch ( Exception $exception ) {
+			$this->logger->error( '[AXO] 3DS return processing failed: ' . $exception->getMessage() );
+			return array(
+				'result'  => 'failure',
+				'message' => $this->get_user_friendly_error_message( $exception ),
+			);
+		}
+
+		WC()->cart->empty_cart();
+
+		return array(
+			'result'   => 'success',
+			'redirect' => $this->get_return_url( $wc_order ),
+		);
+	}
+
+	/**
+	 * Convert exceptions to user-friendly messages.
+	 */
+	private function get_user_friendly_error_message( Exception $exception ) {
+		$error_message = $exception->getMessage();
+
+		// Handle specific error types with user-friendly messages.
+		if ( $exception instanceof DomainException ) {
+			if ( strpos( $error_message, 'Could not capture' ) !== false ) {
+				return __( '3D Secure authentication was unavailable or failed. Please try a different payment method or contact your bank.', 'woocommerce-paypal-payments' );
+			}
+		}
+
+		if ( strpos( $error_message, 'declined' ) !== false ||
+			strpos( $error_message, 'PAYMENT_DENIED' ) !== false ||
+			strpos( $error_message, 'INSTRUMENT_DECLINED' ) !== false ||
+			strpos( $error_message, 'Payment provider declined' ) !== false ) {
+			return __( 'Your payment was declined after 3D Secure verification. Please try a different payment method or contact your bank.', 'woocommerce-paypal-payments' );
+		}
+
+		if ( strpos( $error_message, 'session' ) !== false ||
+			strpos( $error_message, 'expired' ) !== false ) {
+			return __( 'Payment session expired. Please try your payment again.', 'woocommerce-paypal-payments' );
+		}
+
+		return __( 'There was an error processing your payment. Please try again or contact support.', 'woocommerce-paypal-payments' );
+	}
+
+	/**
+	 * Extract payer action URL from PayPal order.
+	 */
+	private function get_payer_action_url( Order $order ) {
```

Comment:
```suggestion
	private function get_payer_action_url( Order $order ): string {
```

Forgive me if I am wrong, but shouldn't this method return string?